From 1ca7ff2db879bb04314fe78f2db7fdafe02e7408 Mon Sep 17 00:00:00 2001
From: ywan171 <yi.a.wang@intel.com>
Date: Thu, 19 Jul 2012 19:07:31 +0800
Subject: [PATCH 028/170] [PORT FROM MAIN] libstagefright: enable AVI playback
 on JB

BZ: 47452

original patch:
34232: libstagefright: enable avi playback on ICS
39311: libstagefright: skip user define chunk in index chunk in AVI Extractor
52722: framework: re-enable AVI for CTP
Signed-off-by: ywan171 <yi.a.wang@intel.com>

Change-Id: I4166761d8e1b6848990b510431919c309a1a1e4c
Signed-off-by: Xiaolin Zhang <xiaolin.zhang@intel.com>
Reviewed-on: http://android.intel.com:8080/75096
Reviewed-by: lab_cactus <lab_cactus@intel.com>
Reviewed-by: Yao, Yong <yong.yao@intel.com>
Tested-by: Tong, BoX <box.tong@intel.com>
Reviewed-by: Tong, BoX <box.tong@intel.com>
Reviewed-by: buildbot <buildbot@intel.com>
Tested-by: buildbot <buildbot@intel.com>
---
 media/libstagefright/AVIExtractor.cpp   |   16 +++++++++++++++-
 media/libstagefright/Android.mk         |    1 +
 media/libstagefright/DataSource.cpp     |    2 ++
 media/libstagefright/MediaExtractor.cpp |    3 +++
 4 files changed, 21 insertions(+), 1 deletions(-)

diff --git a/media/libstagefright/AVIExtractor.cpp b/media/libstagefright/AVIExtractor.cpp
index 5a6211e..89d1e9f 100644
--- a/media/libstagefright/AVIExtractor.cpp
+++ b/media/libstagefright/AVIExtractor.cpp
@@ -577,10 +577,17 @@ static const char *GetMIMETypeForHandler(uint32_t handler) {
         case FOURCC('a', 'v', 'c', '1'):
         case FOURCC('d', 'a', 'v', 'c'):
         case FOURCC('x', '2', '6', '4'):
+        case FOURCC('h', '2', '6', '4'):
         case FOURCC('H', '2', '6', '4'):
         case FOURCC('v', 's', 's', 'h'):
             return MEDIA_MIMETYPE_VIDEO_AVC;
 
+        case FOURCC('H', '2', '6', '3'):
+        case FOURCC('h', '2', '6', '3'):
+        case FOURCC('S', '2', '6', '3'):
+        case FOURCC('s', '2', '6', '3'):
+            return MEDIA_MIMETYPE_VIDEO_H263;
+
         default:
             return NULL;
     }
@@ -784,6 +791,13 @@ status_t AVIExtractor::parseIndex(off64_t offset, size_t size) {
         uint8_t hi = chunkType >> 24;
         uint8_t lo = (chunkType >> 16) & 0xff;
 
+        // chunType is "0x37467878" equal '7Fxx', it is user defined chunk, now just skip it;
+        if (chunkType == 0x37467878) {
+            data += 16;
+            size -= 16;
+            continue;
+        }
+
         if (hi < '0' || hi > '9' || lo < '0' || lo > '9') {
             return ERROR_MALFORMED;
         }
@@ -877,7 +891,7 @@ status_t AVIExtractor::parseIndex(off64_t offset, size_t size) {
 
             double avgChunkSize = 0;
             size_t j;
-            for (j = 0; j <= numSamplesToAverage; ++j) {
+            for (j = 0; j < numSamplesToAverage; ++j) {
                 off64_t offset;
                 size_t size;
                 bool isKey;
diff --git a/media/libstagefright/Android.mk b/media/libstagefright/Android.mk
index 187f7a0..3153436 100644
--- a/media/libstagefright/Android.mk
+++ b/media/libstagefright/Android.mk
@@ -11,6 +11,7 @@ LOCAL_SRC_FILES:=                         \
         AMRWriter.cpp                     \
         AudioPlayer.cpp                   \
         AudioSource.cpp                   \
+        AVIExtractor.cpp                  \
         AwesomePlayer.cpp                 \
         CameraSource.cpp                  \
         CameraSourceTimeLapse.cpp         \
diff --git a/media/libstagefright/DataSource.cpp b/media/libstagefright/DataSource.cpp
index 9d0eea2..cff8a7f 100644
--- a/media/libstagefright/DataSource.cpp
+++ b/media/libstagefright/DataSource.cpp
@@ -19,6 +19,7 @@
 #if CHROMIUM_AVAILABLE
 #include "include/chromium_http_stub.h"
 #endif
+#include "include/AVIExtractor.h"
 
 #include "include/AACExtractor.h"
 #include "include/DRMExtractor.h"
@@ -122,6 +123,7 @@ void DataSource::RegisterDefaultSniffers() {
     RegisterSniffer(SniffAAC);
     RegisterSniffer(SniffMPEG2PS);
     RegisterSniffer(SniffWVM);
+    RegisterSniffer(SniffAVI);
 
     char value[PROPERTY_VALUE_MAX];
     if (property_get("drm.service.enabled", value, NULL)
diff --git a/media/libstagefright/MediaExtractor.cpp b/media/libstagefright/MediaExtractor.cpp
index b18c916..6e33984 100644
--- a/media/libstagefright/MediaExtractor.cpp
+++ b/media/libstagefright/MediaExtractor.cpp
@@ -19,6 +19,7 @@
 #include <utils/Log.h>
 
 #include "include/AMRExtractor.h"
+#include "include/AVIExtractor.h"
 #include "include/MP3Extractor.h"
 #include "include/MPEG4Extractor.h"
 #include "include/FragmentedMP4Extractor.h"
@@ -115,6 +116,8 @@ sp<MediaExtractor> MediaExtractor::Create(
         ret = new MatroskaExtractor(source);
     } else if (!strcasecmp(mime, MEDIA_MIMETYPE_CONTAINER_MPEG2TS)) {
         ret = new MPEG2TSExtractor(source);
+    } else if (!strcasecmp(mime, MEDIA_MIMETYPE_CONTAINER_AVI)) {
+        ret = new AVIExtractor(source);
     } else if (!strcasecmp(mime, MEDIA_MIMETYPE_CONTAINER_WVM)) {
         // Return now.  WVExtractor should not have the DrmFlag set in the block below.
         return new WVMExtractor(source);
-- 
1.7.9

